cli/command/container/create.go

```diff
@@ -77,13 +78,19 @@ func pullImage(ctx context.Context, dockerCli command.Cli, image string, out io.
 	}
 
 	// Resolve the Repository name from fqn to RepositoryInfo
-	repoInfo, err := registry.ParseRepositoryInfo(ref)
+	repoInfo, err := dockerregistry.ParseRepositoryInfo(ref)
 	if err != nil {
 		return err
 	}
 
-	authConfig := command.ResolveAuthConfig(ctx, dockerCli, repoInfo.Index)
-	encodedAuth, err := command.EncodeAuthToBase64(authConfig)
+	authConfig, warns, err := registry.ResolveAuthConfig(ctx, dockerCli.Client(), dockerCli.ConfigFile(), repoInfo.Index)
+	for _, w := range warns {
+		fmt.Fprintf(dockerCli.Err(), "Warning: %v\n", w)
+	}
+	if err != nil {
+		return err
+	}
+	encodedAuth, err := registry.EncodeAuthToBase64(authConfig)
```

Comment:
I notice this code seems to be duplicated a bunch of times. I looked through the use of `authConfigs` and it seems that in almost every single case we either immediately encode them, or in a few cases we pass it around a bit, then encode it later on.

I'd like to suggest a refactoring which will remove this duplication. Since we're making this change to expose an interface I think it's probably the right time to fix this, as changing it later will break the interface.

Instead of running these 3 returns values, I propose we return a new `AuthConfig` struct (as noted elsewhere I believe that `err` is unnecessary and can be removed).

In `registry/registry.go`:

```go
type AuthConfig struct {
   config types.AuthConfig
   warnings []string
}

func (a *AuthConfig) PrintWarnings(out io.Writer) {
    ... // the for loop and fmt.Printf()
}

func (s *AuthConfig) Encode() (string, error) {
    return EncodeAuthToBase64(s.config)
}

func (s *AuthConfig) Config() types.AuthConfig {
    return s.config
}
```

Then consuming code could call:

```go
authConfig := registry.ResolveAuthConfig(ctx, dockerCli.Client(), dockerCli.ConfigFile(), repoInfo.Index)
authConfig.PrintWarnings(dockerCli.Err())

encodedAuth, err := authConfig.Encode()
if err != nil {
    return err
}
```

Or it could pass around the `authConfig` and worry about encoding later.

WDYT?